blpapi.ts

```diff
@@ -0,0 +1,728 @@
+                this.removeAllListeners(listenerName);
+                handler(ev.data);
+            };
+
+            this.once('SessionStarted',
+                      listener.bind(this, 'SessionStartupFailure', (data: any): void => {
+                          log('Session started');
+                          trace(data);
+                          resolve();
+                      }));
+
+            this.once('SessionStartupFailure',
+                      listener.bind(this, 'SessionStarted', (data: any): void => {
+                          log('Session start failure');
+                          trace(data);
+                          reject(new BlpApiError(data));
+                      }));
+        }).nodeify(cb);
+    }
+
+    stop(cb?: (err: any, value: any) => void): Promise<void> {
+        return this.stopped = this.stopped ||
+                              new Promise<void>((resolve: Function, reject: Function): void => {
+            log('Stopping session');
+            this.invoke(this.session.stop);
+            this.once('SessionTerminated', (ev: any): void => {
+                resolve();
+            });
+        }).nodeify(cb);
+    }
+
+
+    // 'request' function overloads.
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            callback: IRequestCallback): void;
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            identity: IIdentity,
+            callback: IRequestCallback): void;
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            label: string,
+            callback: IRequestCallback): void;
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            identity: IIdentity,
+            label: string,
+            callback: IRequestCallback): void;
+
+    // 'request' function definition
+    request(uri: string,
+            requestName: string,
+            request: any,
+            arg3: Object | string | IRequestCallback,
+            arg4?: string | IRequestCallback,
+            arg5?: IRequestCallback): void
+    {
+        var args = Array.prototype.slice.call(arguments);
+        // Get callback and remove it from array.
+        var callback = args.splice(args.length - 1)[0];
+        // Get optional identity and label arguments (idx 3 and 4 if present).
+        var optionalArgs = args.slice(3);
+
+        var executeRequest = (correlatorId: number): void => {
+            var args = [this.session.request,
+                        uri,
+                        requestName,
+                        request,
+                        correlatorId].concat(optionalArgs);
+            this.invoke.apply(this, args);
+        };
+        this.doRequest(uri, requestName, request, executeRequest, callback);
+    }
+
+    generateToken(cb?: (err: any, value: any) => void): Promise<string>
```

Comment:
not sure whether we want to expose `generateToken` as the api name. Maybe `authenticate`? It is more consistent with the 2 step process in the developer guide.